Std/Data/Array/Lemmas.lean

```diff
@@ -164,6 +164,23 @@ theorem swapAt!_def (a : Array α) (i : Nat) (v : α) (h : i < a.size) :
     a.pop[i] = a[i]'(Nat.lt_of_lt_of_le (a.size_pop ▸ hi) (Nat.sub_le _ _)) :=
   List.get_dropLast ..
 
+theorem eq_empty_of_size_eq_zero {as : Array α} (h : as.size = 0) : as = #[] := by
+  apply ext
+  · simp [h]
+  · intros; contradiction
+
+theorem eq_push_of_size_ne_zero [Inhabited α] {as : Array α} (h : as.size ≠ 0) :
+    as = as.pop.push as.back := by
```

Comment:
Might be nice to have a version of this with an existential too, so that rewrites don't get confused

Reply:
Do you mean something like this?
```lean
theorem eq_push_of_size_ne_zero' {as : Array α} (h : as.size ≠ 0) : ∃ (bs : Array α) (c : α), as = bs.push c := sorry
```

Reply:
Ah! Right, the `Inhabited` instance will cause rewrite issues.